### DISCUSSION TITLE
Custom blocklist.txt configuration

### DISCUSSION THREAD

**Discussion by squromiv** [TechnitiumSoftware/DnsServer#882] [Q&A]
I use Docker with v. Technitium 11.5.3.
Placed some blocklist.txt into /var/lib/docker/volumes/technitium_config/_data/www
I set  Settings->Blocking->Quick Add->http://localhost:5380/blocklist.txt)
Then I click "Update now" and nothing happens. There is no temp file in /technitium_config/_data/blocklists directory.
Should I take any additional actions in order to force it to work?

**Comment by ShreyasZare** [comment]
Thanks for asking. I would suggest that you check the Logs from the admin web panel first to know why it did not work. That will give you clue to fix it. You can also use something like wget/curl to test the URL from within the container to see if its working.

Also, do update the DNS server to the latest version since there are a lot of bug fixes.

**Comment by squromiv** [comment]
> do update the DNS server to the latest version

Done.

Also I have placed blocklist.txt directly into container by command

`sudo docker cp /var/lib/docker/volumes/technitium_config/_data/www/blocklist.txt 47cdec3e4962:/opt/technitium/dns/www/blocklist.txt`

So the blocklist is accessible now with setting "file:///opt/technitium/dns/www/blocklist.txt" in Allow / Block List URLs and update process is going correctly.

But if I set "http://localhost:5380/blocklist.txt", update freezes and there is some info in log

> [2024-03-18 15:02:33 Local] DNS Server is downloading block list: http://localhost:5380/blocklist.txt
> [2024-03-18 15:02:33 Local] [IP:0] [admin] Block list update was triggered.
> [2024-03-18 15:02:33 Local] DNS Server failed to download block list and will use previously downloaded file (if available): http://localhost:5380/blocklist.txt
> 
> System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
>  ---> System.Security.Authentication.AuthenticationException: The remote certificate is invalid because of errors in the certificate chain: PartialChain
>    at System.Net.Security.SslStream.SendAuthResetSignal(ReadOnlySpan`1 alert, ExceptionDispatchInfo exception)
>    at System.Net.Security.SslStream.CompleteHandshake(SslAuthenticationOptions sslAuthenticationOptions)
>    at System.Net.Security.SslStream.ForceAuthenticationAsync[TIOAdapter](Boolean receiveFirst, Byte[] reAuthenticationData, CancellationToken cancellationToken)
>    at System.Net.Http.ConnectHelper.EstablishSslConnectionAsync(SslClientAuthenticationOptions sslOptions, HttpRequestMessage request, Boolean async, Stream stream, CancellationToken cancellationToken)
>    --- End of inner exception stack trace ---
>    at System.Net.Http.ConnectHelper.EstablishSslConnectionAsync(SslClientAuthenticationOptions sslOptions, HttpRequestMessage request, Boolean async, Stream stream, CancellationToken cancellationToken)
>    at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
>    at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
>    at System.Net.Http.HttpConnectionPool.AddHttp11ConnectionAsync(QueueItem queueItem)
>    at System.Threading.Tasks.TaskCompletionSourceWithCancellation`1.WaitWithCancellationAsync(CancellationToken cancellationToken)
>    at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
>    at System.Net.Http.DecompressionHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
>    at TechnitiumLibrary.Net.Http.Client.HttpClientNetworkHandler.InternalSendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Http\Client\HttpClientNetworkHandler.cs:line 99
>    at TechnitiumLibrary.Net.Http.Client.HttpClientNetworkHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Http\Client\HttpClientNetworkHandler.cs:line 242
>    at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
>    at DnsServerCore.Dns.ZoneManagers.BlockListZoneManager.<>c__DisplayClass20_0.<<UpdateBlockListsAsync>g__DownloadListUrlAsync|0>d.MoveNext() in Z:\Technitium\Projects\DnsServer\DnsServerCore\Dns\ZoneManagers\BlockListZoneManager.cs:line 427

**Reply by ShreyasZare** [reply]
Thanks for the error log. It seems that you have HTTP to HTTPS redirection enabled and also using self-signed SSL certificate which is causing the block list URL to redirect to HTTPS URL and since the certificate is fails validation, its causing this issue.

The localhost URL concept is quite old. Using `file://` URL is better option now.

**Reply by squromiv** [reply]
> HTTP to HTTPS redirection enabled and also using self-signed SSL certificate

Exactly. Thanks for clarification.
But is it possible somehow to ignore this error in order to get a blocklist by http/https address? In some cases it may be useful. As for example, there is such an option in [FreshRSS](https://github.com/FreshRSS/FreshRSS/) app.
![image](https://github.com/TechnitiumSoftware/DnsServer/assets/15985877/42a54682-af5b-4d5d-9629-fa5368343a2f)

Also I`d like to know if there any "easy" method to edit blocklists inside default Technitium Docker container. It looks a bit annoying - to edit self-made blocklist and then to copy it into container every time.

**Reply by ShreyasZare** [reply]
> But is it possible somehow to ignore this error in order to get a blocklist by http/https address?

Yes, its possible to do that in code or with a new option. However, the blocklist.txt file on the admin web server was a very old option to load a locally stored block list. This approach is now not needed since you can directly use `file://` path which works without much config requirements and is recommended. 

> Also I`d like to know if there any "easy" method to edit blocklists inside default Technitium Docker container. It looks a bit annoying - to edit self-made blocklist and then to copy it into container every time.

You can use the Allowed & Blocked tabs on the admin panel instead. Those are intended to be used directly by the user without need to create custom block list files.
